Source/Core/InputCommon/ControllerInterface/CemuHookUDPServer/CemuHookUDPServer.h

```diff
@@ -0,0 +1,81 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <chrono>
+
+#include "Common/Logging/Log.h"
+#include "Common/Matrix.h"
+#include "InputCommon/ControllerInterface/CemuHookUDPServer/CemuHookUDPServerProto.h"
+#include "InputCommon/ControllerInterface/ControllerInterface.h"
+#include "SFML/Network/UdpSocket.hpp"
+
+namespace ciface::CemuHookUDPServer
+{
+void Init();
+void PopulateDevices();
+void DeInit();
+void SaveSettings();
+
+class Device : public Core::Device
+{
+private:
+  template <class T>
+  class Button : public Core::Device::Input
+  {
+  public:
+    Button(std::string name, const T& buttons, unsigned mask)
+        : m_name(name), m_buttons(buttons), m_mask(mask)
+    {
+    }
+    std::string GetName() const override { return m_name; }
+    ControlState GetState() const override { return (m_buttons & m_mask) > 0; }
+
+  private:
+    const std::string m_name;
+    const T& m_buttons;
+    unsigned m_mask;
+  };
+
+  template <class T>
+  class AnalogInput : public Core::Device::Input
+  {
+  public:
+    AnalogInput(std::string name, const T& input, ControlState range, ControlState offset = 0)
+        : m_name(name), m_input(input), m_range(range), m_offset(offset)
+    {
+    }
+    std::string GetName() const override { return m_name; }
+    ControlState GetState() const override { return (ControlState(m_input) + m_offset) / m_range; }
+
+  private:
+    const std::string m_name;
+    const T& m_input;
+    const ControlState m_range;
+    const ControlState m_offset;
+  };
+
+public:
+  void UpdateInput() override;
+
+  Device(Proto::DsModel model, int index);
```

Comment:
This should likely also have the destructor placed within the cpp file, given this is a non-trivial type.

Reply:
I moved the class definition to the cpp file, as you suggested. Since there is no explicit destructor, does that solve this issue?

Reply:
Could you take a look at my question?

Reply:
@lioncash